### DISCUSSION TITLE
Registering $ objects?

### DISCUSSION THREAD

**Discussion by technicallypete** [nuxt/nuxt#18053] [Questions]
So, as an example, Auth0  wants you to do this:

```
export const Auth0Plugin = {
  install(Vue, options) {
    Vue.prototype.$auth = useAuth0(options);
  }
};
```

But in Nuxt3, A) what's the best way to get access to `Vue`? Currently, I'm doing:

```
export default defineNuxtPlugin(({ vueApp, ssrContext }) => {
  const Vue = vueApp._component.constructor
```

Because vite complains when doing `import Vue from 'vue' etc

And B) (with the above strategy currently), setting `Vue.prototype.$auth = useAuth0(options)` (which is the same as `vueApp._component.constructor.prototype.$auth = useAuth0(options)`), i get this error:

```
[Vue warn]: Component "$auth" has already been registered in target app.
strat is not a function
  at mergeOptions (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:2347:31)
  at ./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:2325:39
  at Array.forEach (<anonymous>)
  at resolveMergedOptions (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:2325:26)
  at applyOptions (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:2022:21)
  at finishComponentSetup (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:6390:9)
  at handleSetupResult (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:6335:5)
  at setupStatefulComponent (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:6298:13)
  at setupComponent (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:6232:11)
  at renderComponentVNode (./node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:198:17)
```

So either vueApp._component.constructor is a bad solution, or proper vite external Vue import needs to be figured out, or there is some setter on the prototype?

Or just really, I guess my main question is: **How are people registering these $ helpers in Nuxt3???**

Also, no need to point me to @nuxtjs/auth-next... A) it currently doesn't seem to work (something something middleware file does not exist), and B) i need auth0 solution to work in storybook, not just in nuxt, etc.

Also, the `useAuth0()` fn isn't the issue.

I'm using nuxtjs/auth-next as a reference though for how THEY are registering $auth but kinda lost in the weeds at the moment on that front so far..

**Comment by danielroe** [comment]
In your case, try:

```js
export default defineNuxtPlugin(() => {
  return {
    provide: {
      auth: useAuth0(options)
    }
  }
})
```

This is the preferred option, and comes with type hinting for `$auth`.

Alternatively, you can do:

```js
export default defineNuxtPlugin((nuxtApp, inject) => {
  inject('auth', useAuth0(options))
})
```

(Incidentally, a lot of the Vue 2 instructions will no longer work, as Vue 3 has a different method for installing. It's no longer correct to install global properties on the Vue constructor - which in fact no longer exists - see https://v3.vuejs.org/guide/migration/global-api.html#vue-prototype-replaced-by-config-globalproperties.)
